### ISSUE TITLE
Kafka Consumer - handler concurrency limit?

### ISSUE THREAD

**Issue by scmorse** [tulios/kafkajs#148]
I did not see anything in the documentation related to this. Basically, if I have a consumer for a topic with 100 partitions, and it is the leader for 50 of those partitions, I don't want it to process all 50 partitions in parallel. I might want to limit it to only processing 5 or 10 partitions concurrently, to limit resource usage, etc.

If there isn't such a limit defined, consider this a feature request  :)

**Comment by Nevon** [comment]
Is this something you've experienced as a problem, or are you considering that it could become a problem?

The overhead of one consumer consuming from many partitions should generally be fairly minimal. Each fetch would involve more requests depending on the number of brokers, but all in all I wouldn't consider that a very large cost.

If the problem is that you're processing a whole batch of messages in parallel, so that you're constrained on resources because you're processing 1000 messages at the same time, you can control the number of messages that you get in each batch with `maxBytesPerPartition` and/or `maxBytes` so that each fetch contains fewer messages. You could also in your application code simply process the messages in "mini batches". For example, if you get 1000 messages in the batch, maybe process 100 in parallel at a time.

We are experimenting a bit with creating utilities for simplifying higher level patterns outside of kafkajs itself, so maybe this is a candidate, but my gut feeling is that if you just control the batch size, you already have a good solution. Let me know if I'm missing something.

**Comment by tulios** [comment]
Exactly what @Nevon said, the concurrency will be limited by the size of your cluster. KafkaJS only makes requests to the leaders, so it should be a fraction of your cluster. However, I don't disagree with the feature request, I think it's a nice feature to have. @scmorse, are you having issues with resource usage?

**Comment by scmorse** [comment]
So, the full story is that we used to use the library `no-kafka`, which had a consumer concurrency limit config value. We set it to 10. Now we're upgrading to use `kafkajs` because this lib has support for creating topics, but it is missing a few things that `no-kafka` had, such as the concurrency control (and some group admin functions like `listGroups` / `fetchConsumerLag`). We may be okay even without this, but it makes it a riskier upgrade, because it greatly increases the number of concurrent messages being processed.

We're not processing batches of messages from one partition in parallel, the concern was more about processing messages from many different partitions in parallel. If I have a topic with 100 partitions for example, then between all of my consumers I'd have 100 messages being processed concurrently, which could put significant strain on a database or other resources. Whereas with a concurrency limit of 10 on each of my 2 consumers, I'd only have at most 20 messages being processed concurrently.
